core/new-gui/src/app/workspace/component/navigation/navigation.component.ts

```diff
@@ -261,35 +273,38 @@ export class NavigationComponent implements OnInit {
   }
 
   /**
-   * Returns true if there's any operator on the graph; false otherwise
+   * Persist the cachedWorkflow in the localStorage, save it to the backend database. It also fetches the new workflow.wid
+   * and updates it in the cache if this is a new workflow.
    */
-  public hasOperators(): boolean {
-    return this.workflowActionService.getTexeraGraph().getAllOperators().length > 0;
-  }
-
-  public onClickSaveWorkflow(): void {
-    if (!this.userService.isLogin()) {
-      alert('please login');
-    } else {
-      const cachedWorkflow: Workflow | null = this.cachedWorkflowService.getCachedWorkflow();
-      if (cachedWorkflow != null) {
-        this.isSaving = true;
-        this.workflowPersistService.persistWorkflow(cachedWorkflow).subscribe(this.cachedWorkflowService.cacheWorkflow).add(() => {
-          this.isSaving = false;
-        });
-      } else {
-        alert('No workflow found in cache.');
-      }
+  public persistCachedWorkflow(): void {
+    const cachedWorkflow: Workflow | undefined = this.workflowCacheService.getCachedWorkflow();
+    if (cachedWorkflow !== undefined) {
+      this.isSaving = true;
+      this.workflowPersistService.persistWorkflow(cachedWorkflow).subscribe((updatedWorkflow: Workflow) => {
+        this.workflowCacheService.cacheWorkflow(updatedWorkflow);
+        this.isSaving = false;
+      });
     }
   }
 
+  /**
+   * Handler for changing workflow name input box, updates the cachedWorkflow and persist to database.
+   */
   onWorkflowNameChange() {
-    this.cachedWorkflowService.setCachedWorkflowName(this.currentWorkflowName);
+    this.workflowCacheService.setCachedWorkflowName(this.currentWorkflowName);
```

Comment:
Just want to make sure in this implementation, everything has a "single source of truth". In your design, it seems like "cache workflow service" has is the single source of truth.

Reply:
Yes, `WorkflowCacheService` is the single source of truth. all the workspace events are changing the cached workflow. And WorkflowPersistService persists the cached workflow to the backend.